tokio/src/task/local.rs

```diff
@@ -0,0 +1,630 @@
+    ///
+    /// # Notes
+    ///
+    /// Since this function internally calls [`Runtime::block_on`], and drives
+    /// futures in the local task set inside that call to `block_on`, the local
+    /// futures may not use [in-place blocking]. If a blocking call needs to be
+    /// issued from a local task, the [`spawn_blocking`] API may be used instead.
+    ///
+    /// For example, this will panic:
+    /// ```should_panic
+    /// use tokio::runtime::Runtime;
+    /// use tokio::{task, blocking};
+    ///
+    /// let mut rt = Runtime::new().unwrap();
+    /// let local = task::LocalSet::new();
+    /// local.block_on(&mut rt, async {
+    ///     let join = task::spawn_local(async {
+    ///         let blocking_result = blocking::in_place(|| {
+    ///             // ...
+    ///         });
+    ///         // ...
+    ///     });
+    ///     join.await.unwrap();
+    /// })
+    /// ```
+    /// This, however, will not panic:
+    /// ```
+    /// use tokio::runtime::Runtime;
+    /// use tokio::{task, blocking};
+    ///
+    /// let mut rt = Runtime::new().unwrap();
+    /// let local = task::LocalSet::new();
+    /// local.block_on(&mut rt, async {
+    ///     let join = task::spawn_local(async {
+    ///         let blocking_result = blocking::spawn_blocking(|| {
+    ///             // ...
+    ///         }).await;
+    ///         // ...
+    ///     });
+    ///     join.await.unwrap();
+    /// })
+    /// ```
+    ///
+    /// [`spawn_local`]: fn.spawn_local.html
+    /// [`Runtime::block_on`]: ../struct.Runtime.html#method.block_on
+    /// [in-place blocking]: ../blocking/fn.in_place.html
+    /// [`spawn_blocking`]: ../blocking/fn.spawn_blocking.html
+    pub fn block_on<F>(&self, rt: &mut crate::runtime::Runtime, future: F) -> F::Output
+    where
+        F: Future + 'static,
+        F::Output: 'static,
+    {
+        let scheduler = self.scheduler.clone();
+        self.scheduler
+            .with(move || rt.block_on(LocalFuture { scheduler, future }))
+    }
+}
+
+impl Default for LocalSet {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<F: Future> Future for LocalFuture<F> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let scheduler = this.scheduler;
+        let mut future = this.future;
+        scheduler.waker.register_by_ref(cx.waker());
+
+        if let Poll::Ready(output) = future.as_mut().poll(cx) {
+            return Poll::Ready(output);
+        }
+
+        scheduler.tick();
+        Poll::Pending
+    }
+}
+
+// === impl Scheduler ===
+
+impl Schedule for Scheduler {
+    fn bind(&self, task: &Task<Self>) {
+        assert!(self.is_current());
+        unsafe {
+            (*self.tasks.get()).insert(task);
+        }
+    }
+
+    fn release(&self, _: Task<Self>) {
+        unreachable!("tasks should only be completed locally")
```

Comment:
Can the JoinHandle being sent to another thread cause this to be hit?

Reply:
Looking at the source, it seems that it is possible for this to get called as part of cancellation from the join handle, so it does need to be handled. The solution is to use a transfer stack similar to the other runtimes.

Reply:
@carllerche I'm not sure if that's the case. Commit b919c958da87e04eb3548589512a67693d265881 adds a test for sending a local future's `JoinHandle` to another thread and awaiting it, and it doesn't hit this `unreachable!`. Am I missing something?